Adding feedback for reading files

adragnevVW opened the pull request.

AndriiVoitenko commented:
Hi @adragnevVW, thank you for the suggestion, but it would not really solve my issue. My problem was that execution of IGdtfFixture::ReadFromFile could take a big amount of time, I do not get an estimation about it, and do not have any chases to cancel the execution. With your changes I can set up a call back and it will send me some string, but I still could not get any assumption how long execution would take and I can not break the execution. And why call back provides a string? Are you expect, that software which is using this library will send those string to the user? 

""Reading : " + fName + " " + nodeName" maybe useful for debugging purpose, but not for the user. 

I expected to get a count, it could be bytes, lines, or event parts of GDTF file and the a number how much bytes, lines, parts already proceed. 

And again abort flag should be part of the call back. then software, which is using library can abort the execution (user do not want to wait, not enough memory for creation of new objects...)

adragnevVW commented:
Hi @AndriiVoitenko,
So what I am understanding you want only to get the file size, how much lines is the actual file and how much lines are already read? Also to add a flag that can cancel the file reading.

AndriiVoitenko commented:
Again, for progress bar I need total count and currently proceeded count. And you can choose what it is better for you to count: lines, or bytes, or GDTF parts. 

Logically total count and proceed count should have same unit. 

And yes, Abort flag is also needed

adragnevVW commented:
Hi, @AndriiVoitenko 
I updated the review adding a new way of feedback and also adding abort flag callback.

#Feedback callback:
I have updated the callback and now we are returning the nodes that fixture contains and we are counted every read node. Also the file name.

#Abort Flag:
I have create an abort flag callback, which is global and once you have canceled the operation it will stop until new file is opened for read.

I've added also a video with example of how it works in VW.


https://github.com/user-attachments/assets/f91426cd-f14e-4b20-878a-10d4b74a1297

ILane-VW commented:
Andrii says that from a code standpoint these changes look good and can be merged in. He has not had a chance to implement the changes yet, but the code is approved. There might be a follow up to this if an issue is found in the future.

We should also utilize this functionality in the fixture builder to provide a progress bar and a cancel operation when opening a gdtf file in the fixture builder. I will create a separate issue for that so that we can make a future change.

AndriiVoitenko commented:
@adragnevVW sorry to say it, but I overlooked, that you build in SetCallBack into a CMediaRessourceVectorImpl. Those class used only for MVR import and I requested a Feedback for GDTF Import, so it should be in a IGdtfFixture, same as SetAbortCallback.

For MVR we already have GetSceneObjectCount to get whole amount of objects in MVR file and objects are proceed one by one, so I can count them by my self.
